env/kube_registry.go

```diff
@@ -0,0 +1,331 @@
+	}
+
+	if err = out.RegisterResources(); err != nil {
+		return nil, err
+	}
+
+	return out, nil
+}
+
+// RegisterResources creates third party resources
+func (kr *KubernetesRegistry) RegisterResources() error {
+	var out error
+	for kind, v := range kr.mapping {
+		apiName := KindToAPIName(kind)
+		// initialize resource if it does not exist
+		res, err := kr.client.Extensions().ThirdPartyResources().Get(apiName)
+		if err != nil {
+			if errors.IsNotFound(err) {
+				tpr := &v1beta1.ThirdPartyResource{
+					ObjectMeta: v1.ObjectMeta{
+						Name: apiName,
+					},
+					Versions: []v1beta1.APIVersion{
+						{Name: IstioResourceVersion},
+					},
+					Description: v.Description,
+				}
+				_, err = kr.client.Extensions().ThirdPartyResources().Create(tpr)
+				if err != nil {
+					out = multierror.Append(out, err)
+				}
+			} else {
+				out = multierror.Append(out, err)
+			}
+		} else {
+			log.Printf("Resource already exists: %s", res.Name)
+		}
+	}
+	return out
+}
+
+// DeregisterResources removes third party resources
+func (kr *KubernetesRegistry) DeregisterResources() error {
+	var out error
+	for kind := range kr.mapping {
+		apiName := KindToAPIName(kind)
+		err := kr.client.Extensions().ThirdPartyResources().Delete(apiName, &api.DeleteOptions{})
+		if err != nil {
+			out = multierror.Append(out, err)
+		}
+	}
+	return out
+}
+
+// CamelCaseToKabobCase converts "MyName" to "my-name"
+func CamelCaseToKabobCase(s string) string {
+	var out bytes.Buffer
+	for i := range s {
+		if 'A' <= s[i] && s[i] <= 'Z' {
+			if i > 0 {
+				out.WriteByte('-')
+			}
+			out.WriteByte(s[i] - 'A' + 'a')
+		} else {
+			out.WriteByte(s[i])
+		}
+	}
+	return out.String()
+}
+
+// KindToAPIName converts Kind name to 3rd party API group
+func KindToAPIName(s string) string {
+	return CamelCaseToKabobCase(s) + "." + IstioAPIGroup
+}
+
+func (kr *KubernetesRegistry) Get(key model.ConfigKey) (*model.Config, bool) {
+	config := &Config{}
+	err := kr.dyn.Get().
+		Namespace(kr.Namespace).
+		Resource(key.Kind + "s").
+		Name(encodeName(key)).
+		Do().Into(config)
+	if err != nil {
+		log.Printf(err.Error())
+		return nil, false
+	}
+	out, err := kr.convert(key.Kind, config)
+	if err != nil {
+		log.Printf(err.Error())
+		return nil, false
+	}
+	return out, true
+}
+
+func (kr *KubernetesRegistry) Put(obj model.Config) error {
+	if err := kr.mapping.ValidateConfig(obj); err != nil {
+		return err
+	}
+	pb, _ := obj.Content.(proto.Message)
+
+	// Marshal from proto to json bytes
+	m := jsonpb.Marshaler{}
+	bytes, err := m.MarshalToString(pb)
+	if err != nil {
+		return err
+	}
+
+	// Unmarshal from bytes to go map
+	var data map[string]interface{}
+	err = json.Unmarshal([]byte(bytes), &data)
+	if err != nil {
+		return err
+	}
+
+	out := &Config{
+		Metadata: api.ObjectMeta{Name: encodeName(obj.ConfigKey)},
+		Data:     data,
+	}
+
+	err = kr.dyn.Post().
+		Namespace(kr.Namespace).
+		Resource(obj.Kind + "s").
+		Body(out).
+		Do().Error()
+	if err != nil {
```

Comment:
The intent is that err will be consumed by a log service.
It cannot be just returned to the client since that would leak Kubernetes details.